### ISSUE TITLE
package/apt.rb doesn't support epoch in version

### ISSUE THREAD

**Issue by Aketzu** [chef/chef#6606]
## Description

If a package version contains epoch (i.e. starts with `1:`) version_compare will feed it to Gem::Version.new which barfs.

## Chef Version

chef-client -v
Chef: 13.5.25

## Platform Version

Debian 9.2

## Replication Case

Following cookbook recipe and then just waiting for a new version of the package
```
    apt_repository 'zabbix' do
      distribution 'stretch'
      uri 'http://repo.zabbix.com/zabbix/3.4/debian'
      key '082AB56BA14FE591'
      keyserver 'keyserver.ubuntu.com'
      components ['main']
      notifies :run, 'execute[apt-get update]', :immediately
    end

    apt_package 'zabbix-agent' do
      options '-o Dpkg::Options::="--force-confold" -y'
      action :upgrade
    end
```

## Client Output

```
Chef run failed on server.example.com
apt_package[zabbix-agent] (server::zabbix line 75) had an error: ArgumentError: Malformed version number string 1:3.4.4-3-stretch

Backtrace
/opt/chef/embedded/lib/ruby/2.4.0/rubygems/version.rb:208:in `initialize'
/opt/chef/embedded/lib/ruby/2.4.0/rubygems/version.rb:199:in `new'
/opt/chef/embedded/lib/ruby/2.4.0/rubygems/version.rb:199:in `new'
/opt/chef/embedded/lib/ruby/gems/2.4.0/gems/chef-13.5.25/lib/chef/provider/package/apt.rb:134:in `version_compare'
```

Rest in https://gist.github.com/Aketzu/3e6d84fd3f7700cdf0c556bec1d51e9c

## Analysis

package/apt.rb is using Gem::Version for checking which version is never but Gem version numbers follow different format than Debian package versions. Current version numbering is defined in https://www.debian.org/doc/debian-policy/#s-f-version which also defines proper way of comparing version numbers.

https://github.com/chef/chef/blob/913ea3f7f8e3363caa71db6b4e16cbf2ed481363/lib/chef/provider/package/apt.rb#L130-L135
apt.rb version_compare should be changed to use much more complex version comparision logic

Other options are to call system `dpkg --compare-versions` or write a library which uses apt-pkg (e.g. https://anonscm.debian.org/cgit/collab-maint/ruby-debian.git/tree/ext/debian_version/Version.cpp ). Quick search didn't turn up any ruby implementations of Debian version comparision logic.

**Comment by tas50** [comment]
@Aketzu Can you give this a try in 13.6.4. We stopped using Gem to compare versions there so this should be resolved already.

**Comment by Aketzu** [comment]
Source code looks much better there. I thought I was looking at master but instead I was in 913ea3f. I searched for 'package.rb' in github, went to lib/chef/provider/package/windows.rb which pointed to 913ea3f instead of master. From there I chose apt.rb but it still stuck to 913ea3f and I didn't notice that.

Testing is a bit hard due to #6485. I have to compile my own version with proper libssl. I'll do that in a few days.

Sorry for the noise
